src/rust/ensogl/src/display/shape/text/text_field/cursor.rs

```diff
@@ -462,20 +478,27 @@ impl Cursors {
     ///
     /// If after this operation some of the cursors occupies the same position, or their selected
     /// area overlap, they are irreversibly merged.
-    pub fn navigate_all_cursors(&mut self, navigation:&mut CursorNavigation, step:Step) {
+    pub fn navigate_all_cursors(&mut self, navigation:&mut CursorNavigation, step:Step) -> f32 {
         self.navigate_cursors(navigation,step,|_| true)
     }
 
     /// Do the navigation step of all cursors satisfying given predicate.
     ///
     /// If after this operation some of the cursors occupies the same position, or their selected
     /// area overlap, they are irreversibly merged.
-    pub fn navigate_cursors<Predicate>
-    (&mut self, navigation:&mut CursorNavigation, step:Step, mut predicate:Predicate)
-    where Predicate : FnMut(&Cursor) -> bool {
+    pub fn navigate_cursors<P>
+    (&mut self, navigation:&mut CursorNavigation, step:Step, mut predicate:P) -> f32
+    where P : FnMut(&Cursor) -> bool {
         let filtered = self.cursors.iter_mut().filter(|c| predicate(c));
         filtered.for_each(|cursor| navigation.move_cursor(cursor, step));
+        let new_position = self.last_cursor().position;
         self.merge_overlapping_cursors();
+
+        let scroll      = navigation.scroll_position.y;
+        let line_height = navigation.content.line_height;
+        let height      = ((navigation.text_field_size.y - line_height) / line_height).floor() * line_height;
+        let position    = new_position.line as f32 * navigation.content.line_height - scroll;
+        -position.min(position.max(height) - height)
```

Comment:
I don't see why we compute the value, and why this function returns anything. The docs also don't mention that. And also I see we don't use this returned value anywhere. Probably this is change to be reverted.

Reply:
The responsibility of scrolling the text field was moved from CursorNavigation to TextField. Scrolling now happens when navigating and when typing a new line.